core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -196,30 +199,47 @@ public ListBucketResult call() throws S3Exception {
   }
 
   /**
-   * @summary creates an object without using multipart upload
+   * @summary uploads an object or part of an object in multipart upload
    * @param contentMD5 the optional Base64 encoded 128-bit MD5 digest of the object
    * @param bucket the bucket name
    * @param object the object name
+   * @param partNumber the identification of the part of the object in multipart upload,
+   *                   otherwise null
+   * @param uploadId the upload ID of the multipart upload, otherwise null
    * @param is the request body
    * @return the response object
    */
   @PUT
   @Path(OBJECT_PARAM)
   @ReturnType("java.lang.Void")
   @Consumes(MediaType.APPLICATION_OCTET_STREAM)
-  public Response createObject(@HeaderParam("Content-MD5") final String contentMD5,
+  public Response createObjectOrUploadPart(@HeaderParam("Content-MD5") final String contentMD5,
       @PathParam("bucket") final String bucket,
       @PathParam("object") final String object,
+      @QueryParam("partNumber") final Integer partNumber,
+      @QueryParam("uploadId") final Long uploadId,
       final InputStream is) {
     return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response>() {
       @Override
       public Response call() throws S3Exception {
         Preconditions.checkNotNull(bucket, "required 'bucket' parameter is missing");
         Preconditions.checkNotNull(object, "required 'object' parameter is missing");
+        Preconditions.checkArgument((partNumber == null && uploadId == null)
+            || (partNumber != null && uploadId != null),
+            "'partNumber' and 'uploadId' parameter should appear together or be "
+            + "missing together.");
 
         String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
         checkBucketIsAlluxioDirectory(bucketPath);
+
         String objectPath = bucketPath + AlluxioURI.SEPARATOR + object;
+        if (partNumber != null) {
+          // This object is part of a multipart upload, should be uploaded into the temporary
+          // directory first.
+          String tmpDir = S3RestUtils.getMultipartTemporaryDirForObject(bucketPath, object);
```

Comment:
maybe not for this PR, but is there a way for us to garbage collect abandoned multi-part uploads?. also please leave a comment if it makes sense

Reply:
according to the S3 specification, if the multipart upload is neither completed or aborted, the uploaded parts remain in the system.

Reply:
https://stackoverflow.com/questions/29911573/does-aws-s3-automatically-abort-multipart-uploads-after-a-timeout. Seems like may be able to configure a timeout.

Reply:
thanks, left a TODO